### DISCUSSION TITLE
DaFoam Questions

### DISCUSSION THREAD

**Discussion by Janfoam** [mdolab/dafoam#70] [General]
Hi Ping He, 

1) For the wind turbine tutorial, I noticed that no constraints were applied. Does that mean that only unconstrained simulations can be done in this particular case? 

2) "The solveCL computes the angle of attack to match the target_CL by first computing dCL/dAlpha (alpha is the angle of attack). Then, it changes alpha using this formulation: Alpha_New = Alpha_0 + dCL/dAlpha * (CL_target - CL_0). We need to repeat this process until we get good CL_target." The changing of the alpha with the formulation Alpha_New = Alpha_0 + dCL/dAlpha * (CL_target - CL_0) - to be sure Alpha_0 is initial angle of attack it calculated while CL_0 is also the initial angle lift coefficient? Can you please clarify the reasoning behind using the formulation?

Thanks

**Comment by friedenhe** [comment]
1. Yes, we do not have any constraints for the wind turbine case at this moment. However, you are welcome to add thickness or volume constraints, refer to the Onera_M6 case on how to add these geometry constraints.

2. Yes, CL_0 is the initial lift, alpha_0 is the initial angle of attack. The solveCL function assumes the CL vs alpha relationship is linear. In that formulation, dCL/dAlpha is the gradient of CL wrt alpha, computed by using the finite-difference method. Once you know the gradient of the linear CL-alpha function, you can predict CL at any alpha. We need multiple iterations because the CL-alpha relationship is not very linear if CL_0 is far away from CL_target.

**Reply by Janfoam** [reply]
Thanks for your answers Friedenhe. 

I looked through the code for the NREL6 and have some more questions: 

- Why is symmetry planes specified as: symmetryPlanes": []? 
- No CL value is specified at the start( usually global parameters). What is it using for the commands for Task to arrive at a CL?
 **_elif args.task == "solveCL":
    optFuncs.solveCL(CL_target, "alpha", "CL")_**
- There is no genAirFoilMesh.py in the case folder. Is this correct? 

For the Onera6 case
Unfortunately, this aspect is still not clear to me. It is also in the airfoil optimization case. Can you please clarify what you use to define the lines in this case. They could be any lines that are close to the le and te, but how do you ensure that the lines are close to the leading and trailing edge? What is the guiding compass in this case?
# NOTE: the LE and TE lists are not parallel lines anymore, these two lists define lines that
# are close to the leading and trailing edges while being completely within the wing surface
leList = [[0.01, 0.0, 1e-3], [0.7, 0.0, 1.19]]
teList = [[0.79, 0.0, 1e-3], [1.135, 0.0, 1.19]]

**Reply by Janfoam** [reply]
The  large words in bold are coming directly from the code. Not sure why they became so large and bold - this was unintentional :-)

**Reply by friedenhe** [reply]
@Janfoam There is no symmetry plane for the NREL6 case. The objective function is torque. We don't use the genAirFoilMesh.py to generate mesh, instead, we use snappyHexMesh. Read [this page](https://mdolab-mach-aero.readthedocs-hosted.com/en/latest/machAeroTutorials/opt_aero.html) on how to setup leList and teList.

**Reply by Janfoam** [reply]
@friedenhe, Thanks for your answers. I'll be testing the simulations and will soon be in touch with other questions. Enjoy your Holidays.

**Comment by Janfoam** [comment]
Hi He, 

-Can you please elaborate how the boundary conditions are defined in DAFoam. I've haven't come across any explanation for this in any of the publications associated with DAFOAM. 

- How does DAFoam consider the velocity components when varying the angle of attack? 

- For the FFD points, is there a criteria in DAFoam that should be respected?

- For thickness changes- is there an output file to look at to see how the thickness varies after the simulation? From the Constraint setup in DAFoam, the thickness constraint depends on the le and te, which are vary depending on how the user established the lines used to define the te and le. 

Thanks

**Reply by friedenhe** [reply]
1. Check line 127 from [here](https://github.com/DAFoam/tutorials/blob/master/NACA0012_Airfoil/incompressible/runScript.py). We defined the x and y components of velocity at far-field based on the alpha value.
2. The only requirement is that the FFD points should be close to the design surface.
3. The constraint values will be printed to the screen after each flow simulation. Check [this]( https://dafoam.github.io/mydoc_get_started_runscript.html#optimization-progress-logopttxt).

**Comment by Janfoam** [comment]
Thanks

For the boundary conditions I meant how is this determined mathematically? I am comparing DAFoam's discrete adjoint mathematical statements with the continuous adjoint. In the latter, there is a step by step guide as to how to arrive at the boundary conditions, however, I have not yet found that for DAFoam. Can you please clarify? 

Also, is there any possibility to have online workshops/tutoring with DAFoam and if yes who should I contact for this?

**Reply by friedenhe** [reply]
In the discrete adjoint, we don't need special treatment or derive extra formulations for boundary conditions, as opposed to the continuous adjoint. The only thing we need to do is to update the boundary values before evaluating the residual. Check our [paper](https://www.researchgate.net/publication/333518918_Effective_Adjoint_Approaches_for_Computational_Fluid_Dynamics) for details. We do have plans to host workshops, maybe sometime later this year. We will post an announcement when the time comes. Stay tuned.

**Comment by Janfoam** [comment]
Hi, 
I set up a simulation in DAFoam to simulate an airfoil with the aim of seeing how the thickness parameters work. I don't know if this is normal, but it took DAFoam around 4 hours to calculate the angle of attack that I will use for the simulation. The simulation is running, but its somewhere close to 17 hours already. I think it will take maybe another 10 or so to finish. Is this usual? 

For the thickness constraints I noticed that both the trailing edge and leading edge are coded to be fixed in DAFoam and it is not recommended to change this because of the symmetry constraint. However, for a wind turbine while the trailing edge is usually modelled fixed the leading edge is usually free. My question is what can be done in DAFoam for this seeing there is a restriction in the thickness conditions?

How does the FFD points and movement affect the thickness in DAFoam?

**Reply by friedenhe** [reply]
The runtime sounds very unusual. Please attach the log file and runScript.py. Again, check this [page](https://mdolab-mach-aero.readthedocs-hosted.com/en/latest/machAeroTutorials/opt_aero.html) on why we want to fix the LE and TE (it is not because of the symmetry constraint). That page also explains how thickness constraint is implemented.

**Reply by Janfoam** [reply]
Thank you. I'll send the files to you. I don't see an attachment option to attach files.

**Reply by friedenhe** [reply]
I thought you meant the flow simulation took 4 hours, but it seems from the logOpt.txt file that the optimization was going well and the speed was normal (you had 200K cells and ran it with one CPU core). You did have a decent amount of drag reduction. Refer to this [page](https://dafoam.github.io/mydoc_get_started_runscript.html#optimization-progress-logopttxt) on how to interpret the optimization progress. One more thing: your initial lift is not 0.65. You need to run solveCL and get a correct alpha for the target lift, before the optimization. Bottom line, don't start the optimization with an infeasible design.
[logOpt.txt](https://github.com/mdolab/dafoam/files/5822628/logOpt.txt)

**Reply by Janfoam** [reply]
Thank you for your response @friedenhe. I did **run solveCL** with the input CL as 0.65 in the **runScript.py**. This produced an angle of attack of  alpha0 = 6.016734. It took around 4 hrs to produce the current alpha which is being used in the optimization. After that I ran **mpirun -np 4 python runScript.py 2>&1 | tee logOpt.txt** and its currently its 48hrs and still running the optimization. I don't understand why then the initial lift is not 0.65, because this is the lift being used in the optimization. Can you please clarify?

**Reply by friedenhe** [reply]
solveCL log file, please.

**Reply by Janfoam** [reply]
Hi @friedenhe, Unfortunately, there is no solveCL log file... I was trying to copy the info on the terminal, but the copy does not capture the start of the simulation. Any suggestion you may have to get around this is very welcome. 

Some points: 

My **CL_target = 0.65** - this is lift I used to run solveCL. From what I understand so far, these values should be configured in the runScript.py then run **mpirun -np 4 python runScript.py --task=solveCL** and then run **mpirun -np 4 python runScript.py 2>&1 | tee logOpt.txt**. 

However, from the first evaluation of the objective function these values appear: CD': 0.014040335310449507, 'CL': 0.5285905822520857. The simulation ends after some 52hrs and produces these values: CD': 0.012313447669046119, 'CL': 0.6497255165906527.  The target lift is reached and the drag is reduced. However, where is the CL 0.5285905822520857 from the start of the simulation coming from?



   Constraints (i - inequality, e - equality)
      Index  Name                              Type          Lower           Value           Upper    Status  Lagrange Multiplier (N/A)
          0  DVCon1_volume_constraint_0           i   1.000000E+00    0.000000E+00    3.000000E+00         L    9.00000E+100
          1  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
          2  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
          3  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
          4  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
          5  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
          6  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
          7  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
          8  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
          9  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
         10  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
         11  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
         12  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
         13  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
         14  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
         15  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
         16  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
         17  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
         18  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
         19  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
         20  DVCon1_thickness_constraints_0       i   8.000000E-01    0.000000E+00    3.000000E+00         L    9.00000E+100
         21  DVCon1_linear_constraint_0_shapey    e   0.000000E+00    0.000000E+00    0.000000E+00              9.00000E+100
         22  DVCon1_linear_constraint_0_shapey    e   0.000000E+00    0.000000E+00    0.000000E+00              9.00000E+100
         23  DVCon1_linear_constraint_0_shapey    e   0.000000E+00    0.000000E+00    0.000000E+00              9.00000E+100
         24  DVCon1_linear_constraint_0_shapey    e   0.000000E+00    0.000000E+00    0.000000E+00              9.00000E+100
         25  DVCon1_linear_constraint_0_shapey    e   0.000000E+00    0.000000E+00    0.000000E+00              9.00000E+100
         26  DVCon1_linear_constraint_0_shapey    e   0.000000E+00    0.000000E+00    0.000000E+00              9.00000E+100
         27  DVCon1_linear_constraint_0_shapey    e   0.000000E+00    0.000000E+00    0.000000E+00              9.00000E+100
         28  DVCon1_linear_constraint_0_shapey    e   0.000000E+00    0.000000E+00    0.000000E+00              9.00000E+100
         29  DVCon1_linear_constraint_0_shapey    e   0.000000E+00    0.000000E+00    0.000000E+00              9.00000E+100
         30  DVCon1_linear_constraint_0_shapey    e   0.000000E+00    0.000000E+00    0.000000E+00              9.00000E+100
         31  DVCon1_linear_constraint_1_shapey    e   0.000000E+00    0.000000E+00    0.000000E+00              9.00000E+100
         32  DVCon1_linear_constraint_1_shapey    e   0.000000E+00    0.000000E+00    0.000000E+00              9.00000E+100
         **33  CL                                   e   6.500000E-01    0.000000E+00    6.500000E-01         E    9.00000E+100**

![image](https://user-images.githubusercontent.com/76228791/104815328-159af880-5814-11eb-85f7-f134f3e9bfa0.png)

**Reply by friedenhe** [reply]
Run this `mpirun -np 4 python runScript.py --task=solveCL 2>&1 | tee logSolveCL.txt` It is very likely solveCL did not work because you had a not-tight-enough primal convergence tolerance.

**Reply by friedenhe** [reply]
> However, where is the CL 0.5285905822520857 from the start of the simulation coming from?
> 
> Constraints (i - inequality, e - equality)
> Index Name Type Lower Value Upper Status Lagrange Multiplier (N/A)
> 0 DVCon1_volume_constraint_0 i 1.000000E+00 0.000000E+00 3.000000E+00 L 9.00000E+100
> 1 DVCon1_thickness_constraints_0 i 8.000000E-01 0.000000E+00 3.000000E+00 L 9.00000E+100
> ...
> 30 DVCon1_linear_constraint_0_shapey e 0.000000E+00 0.000000E+00 0.000000E+00 9.00000E+100
> 31 DVCon1_linear_constraint_1_shapey e 0.000000E+00 0.000000E+00 0.000000E+00 9.00000E+100
> 32 DVCon1_linear_constraint_1_shapey e 0.000000E+00 0.000000E+00 0.000000E+00 9.00000E+100
> **33 CL e 6.500000E-01 0.000000E+00 6.500000E-01 E 9.00000E+100**
> 

BTW: what is printed here (from pyoptsparse) is what you want for the lift, but your first CFD simulation returns CL 0.52.... You need to adjust alpha to make sure your first CFD simulation has CL=0.65.

**Reply by Janfoam** [reply]
Thanks for the tip @friedenhe . I used the command mpirun -np 4 python runScript.py --task=solveCL 2>&1 | tee logSolveCL.txt as suggested and now I am getting an error. I tried rerunning the case a few times but the error appears. This is the first time I am seeing this. Do you have any clue what this means? 

Alpha was adjusted to alpha0 = 6.016734 after running the solve CL. So this alpha is what was produced after the solving for CL and then the first simulation gave  CL 0.52... despite having the input value for CL_target = 0.65

![image](https://user-images.githubusercontent.com/76228791/104819215-5868ca80-582c-11eb-805a-69986dd5ee76.png)

**Reply by friedenhe** [reply]
Are you sure you loaded the OpenFOAM environment before running this?

**Reply by Janfoam** [reply]
Yes, the openFOAM environment is loaded prior to all optimization attempts. 

I ran this command mpirun -np 4 simpleFoam -parallel to see what is happening with the native solver and got this message below: 

trapFpe: Floating point exception trapping enabled (FOAM_SIGFPE).
fileModificationChecking : Monitoring run-time modified files using timeStampMaster (fileModificationSkew 10)
allowSystemOperations : Allowing user-supplied system call operations

// * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * //
[0] 
[0] 
[0] --> FOAM FATAL ERROR: 
[0] simpleFoam: cannot open case directory "/data/j/Correctmeshcopy/dafoam-2.1.3/tutorials-master/NACA0012_Airfoil/incompressible/processor0"
[0] 
[0] 
FOAM parallel run exiting
[0] 
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD 
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
----------------------------------------------------------------

To my knowledge nothing was done to DAFoam or openFoam to reconfigure anything or delete any files. The last simulation exited fine and had no issues. I don't know why this error appears.

Maybe this is a silly question to ask, but I'll ask it anyway: Could the previous simulation that ran for some 50hrs cause this to happen and if so why?

Thank you @friedenhe for your patience, dedication and motivation to be answering questions even on a Saturday.

**Reply by friedenhe** [reply]
If you can not run the native OpenFOAM solver simpleFoam, you can't run DAFoam! Google the error message.

**Reply by Janfoam** [reply]
I know that if the native OpenFOAM solver simpleFoam is not accessible then DAFoam will not run. There were no issues with the native solver nor DAFoam prior to running the simulation for NACA 0012.  It is very unusual that **after** completing the simulation that these issues are occuring.

**Reply by friedenhe** [reply]
I suggest you first learn the basic usage of OpenFOAM from [here](https://www.openfoam.com/documentation/tutorial-guide/), make sure you can run native OpenFOAM, and know how to deal with OpenFOAM-related errors. For example, this one:  "cannot open case directory "/data/j/Correctmeshcopy/dafoam-2.1.3/tutorials-master/NACA0012_Airfoil/incompressible/processor0" is a very basic error in OpenFOAM. Google how to fix it.

**Reply by Janfoam** [reply]
Hi @friedenhe, Thanks for your quick response. 
Some points: 
- I do know the basics of openFOAM 
- The "cannot open case directory" is not that simple to fix as outlined here and could require administrator privileges: https://www.cfd-online.com/Forums/openfoam-solving/134568-foam-fatal-io-error-cannot-find-file.html
-  Even after the native solver loaded the error with DAFoam continues. It would be helpful to know why this happens so that I can move on with my project. If you are able to provide some insight into what is causing this error, this would be greatly appreciated.

- ![image](https://user-images.githubusercontent.com/76228791/104858056-f340d200-591c-11eb-878e-f3e3bad789e7.png)

**Reply by friedenhe** [reply]
You managed to run solveCL before, and now it is not working. It is hard to tell what have you changed that caused this. So I suggest go back to the last working setup (it could be the tutorial case) and gradually change things, see which step breaks. If you can not run solveCL for the tutorial case. Check your DAFoam installation.

**Reply by Janfoam** [reply]
Thank you @friedenhe for your response and suggestions. I'm still waiting for the administrator to have a look at DAFoam installation after doing the necessary checks myself and could not find the problem.

In the meantime, to not lose momentum in learning about DAFoam, I'll ask some additional questions under different heading on the platform. Subject-specific headings will make it easier for people in the future to find what they want instead of scrolling through a long chat. Will come back to this thread once the issue is resolved.
